csds-client/client/client.go

```diff
@@ -1,92 +1,61 @@
 package client
 
 import (
-	csdspb "github.com/envoyproxy/go-control-plane/envoy/service/status/v2"
-	envoy_type_matcher "github.com/envoyproxy/go-control-plane/envoy/type/matcher"
-
 	"context"
 	"crypto/x509"
 	"errors"
-	"flag"
 	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	csdspb_v2 "github.com/envoyproxy/go-control-plane/envoy/service/status/v2"
+	envoy_type_matcher "github.com/envoyproxy/go-control-plane/envoy/type/matcher"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/credentials"
 	"google.golang.org/grpc/credentials/oauth"
 	"google.golang.org/grpc/metadata"
-	"io"
-	"time"
 )
 
-type Flag struct {
-	uri             string
-	platform        string
-	authnMode       string
-	apiVersion      string
-	requestFile     string
-	requestYaml     string
-	jwt             string
-	configFile      string
-	monitorInterval time.Duration
-	visualization   bool
+type ClientOptions struct {
+	Uri             string
+	Platform        string
+	AuthnMode       string
+	ApiVersion      string
+	RequestFile     string
+	RequestYaml     string
+	Jwt             string
+	ConfigFile      string
+	MonitorInterval time.Duration
+	Visualization   bool
 }
 
 type Client struct {
 	clientConn *grpc.ClientConn
-	csdsClient csdspb.ClientStatusDiscoveryServiceClient
+	csdsClient interface{}
```

Comment:
I was thinking that instead of having an interface as a parameter of Client, you could have Client itself being an interface. This way instead of doing case handling of versions within function implementation, you could have separate and clean implementation. Something like

```
type Client interface {
   Run() error
}

// Then you would have an implementation of this interface for each version
type ClientV2 struct {
}

NewV2() (*ClientV2, err)
NewV3() (*ClientV3, err)

```
In the main then you just select which implementation you want to use:

```
var c Client
if (version == "v2") { c = NewV2() }
else if (version == "v3") { c = NewV3() }
// Execute
c.Run()

```
This way you can keep implementation for different versions cleanly separated.

There might be a lot of common code between V2 and future versions implementation. For this, in the future, we can extract common parts into libraries that will be reused by V2, V3, Vn implementations.

Reply:
Thanks for the suggestions! I've changed `Client` to interface and had `ClientV2` to implement it.